### PR TITLE
Add tests for phonenumbers dependency

### PR DESCRIPTION
**Pull request by gherceg**
## Technical Summary

Initially started as part of an upgrade to phonenumbers in https://github.com/dimagi/commcare-hq/pull/31649, but @joxl brought up a good point that this should be done in a separate PR prior to the upgrade. The first commit here was cherry picked from that PR to preserve the relevance of the feedback Joel gave.

Obviously in addition to adding tests, I made some minor refactors. I liked Cory's initial idea to wrap all uses of the phonenumbers API in this file, which was the impetus for changes in 89201c423005e4aec6bcbd993b7508f55aaae41a and c41e902466f15eb421c6ec28ff7638c7a945716d. This makes it easy to see all uses of this dependency at a glance.

### Potential Risks

I removed the `if phone_number:` check in `parse_phone_number` because after inspecting phonenumbers `parse` method, there does not appear to be the possibility to return None from that method. Either an exception is raised, or a non-None object is returned. Will gladly change back if there is a convincing reason to still check for None.

In regards to the `get_country_code_and_national_number` refactor, I confirmed that tests passed in between 8946560d37aac493f2ebc3d6e87642cdd1e5de33 and e5022e7d284350fd5abe6c64b879bf5f6dd0de96. 

### Safety story

Minor refactors with limited changes of behavior. I've highlighted potential risks with justifications for why they are safe above.

### Automated test coverage

Added automated tests in this PR.

### QA Plan

No QA necessary.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by joxl — APPROVED** [review]
Excellent cleanup and refactor work here. Thanks!

File: corehq/apps/sms/phonenumbers_helper.py [hunk 1]
```diff
@@ -17,23 +17,44 @@ def parse_phone_number(number, region=None, failhard=True):
...
-    parsed = parse_phone_number(number, failhard=failhard)
-    if parsed:
-        country_code = parsed.country_code
-        national_number = strip_plus(number)[len(str(country_code)):]
-        return country_code, national_number
-    return None, None
+def get_country_code_and_national_number(number):
+    """
+    :param number: str representing phone number
+    :return: country_code: int, national_number: str
```
Comment (millerdev) [inline corehq/apps/sms/phonenumbers_helper.py:36]: These types listed here are incomplete since this may return `None, None`
```suggestion
    :return: country_code: int, national_number: str or, if the number
    could not be parsed, `None, None`.
```

File: corehq/apps/sms/phonenumbers_helper.py [hunk 2]
```diff
@@ -17,23 +17,44 @@ def parse_phone_number(number, region=None, failhard=True):
...
+    """
+    parsed = parse_phone_number(number, failhard=False)
+    return (parsed.country_code, str(parsed.national_number)) if parsed else (None, None)
+
+
+def country_code_for_region(region):
+    """
+    Simple wrapper around phonenumbers country_code_for_region method
+    :param region: str (examples: 'US', 'IN', 'ZA')
+    :return: int
```
Comment (millerdev) [inline corehq/apps/sms/phonenumbers_helper.py:46]: Does this always return an int, even if the country code was unknown?

**Review by millerdev — APPROVED** [review]
This is excellent!

File: corehq/apps/sms/phonenumbers_helper.py (hunk 2 above)
Comment (gherceg) [reply corehq/apps/sms/phonenumbers_helper.py:47]: At first glance I thought yes. This is the method in phonenumbers:
```
def country_code_for_region(region_code):
    if not _is_valid_region_code(region_code):
        return 0
    return country_code_for_valid_region(region_code)
```
but if for some reason `_is_valid_region_code` has a bug, `country_code_for_valid_region` will raise an exception for an invalid code. Is that something we should plan for?

Comment (millerdev) [reply corehq/apps/sms/phonenumbers_helper.py:47]: We might not need to make any code changes, but the documentation should ideally note that it could raise an exception.

Comment (gherceg) [reply corehq/apps/sms/phonenumbers_helper.py]: See b1b46aab197c59f82208b34d0e0316e878740d95

File: corehq/apps/sms/phonenumbers_helper.py [hunk 3]
```diff
@@ -17,23 +17,46 @@ def parse_phone_number(number, region=None, failhard=True):
...
+    parsed = parse_phone_number(number, failhard=False)
+    return (parsed.country_code, str(parsed.national_number)) if parsed else (None, None)
+
+
+def country_code_for_region(region):
+    """
+    Simple wrapper around phonenumbers country_code_for_region method
+    :param region: str (examples: 'US', 'IN', 'ZA')
+    :return: int, 0 if invalid, country_code if valid
+    NOTE it is possible for this to raise an Exception if phonenumbers region validation has a bug
```
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/apps/sms/phonenumbers_helper.py:48]: They raise a general `Exception` which is certainly surprising, but maybe that is because from their perspective, it should _never_ happen. 

We can't add a test to exercise this code because it is all internal to phonenumbers. They have a list of supported regions to check if a region is valid, and then they have a method to retrieve `PhoneMetadata` for a region that appears distinct from the list of supported regions, hence the chance of an exception.

Based on your suggestion to change from if to because, I'm wondering if we are on slightly different pages and if this comment clears anything up. I wouldn't say that there is a bug in phonenumbers at the moment.

[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/sms/phonenumbers_helper.py:48]: > They raise a general `Exception`

That is terrible, and either an indication of someone not knowing how to use exceptions or it being an exception that is not meant to be caught (which would be surprising).

> We can't add a test to exercise this code because it is all internal to phonenumbers. ... I wouldn't say that there is a bug in phonenumbers at the moment.

I think I'm not understanding. It seems like there would be specific values, if passed to `country_code_for_region` will cause it to raise an exception. Can we not find such a value and write a test that passes that value into `country_code_for_region` and assert that the exception is raised?

[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/apps/sms/phonenumbers_helper.py:48]: Their expected behavior is to determine that a region is invalid before attempting to use it, and return 0 as the country code (not raise an exception). I have a test for this behavior (see `test_county_code_for_region_returns_zero_if_invalid_region`).

[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/apps/sms/phonenumbers_helper.py]: Offlined with Daniel and it turned out to be some miscommunication. Given that the raised exception is only possible if there is a bug in phonenumbers, we are not going to document it on our side. Removed in 9a16ed9eaeb9a946982ab01180175c3ef8637e74.

**Review by millerdev — APPROVED** [review]

**Comment by gherceg** [comment]
9839a5d520ca17c9ca1000a13bb33d0c4ef605a7 is a big one! I remembered that I looked at `test_gateway_fees.py` last year, and made the note that tests take a long time, and the setup of many random inputs was likely unnecessary. I refactored tests in `test_gateway_charges.py` to include covering areas that `test_gateway_fees.py` covered, as well as the edge case that caused the failing test in this PR.

As described in `test_phone_number_edge_case` (open to better names), the failing test in this PR was due to how the phonenumbers lib was parsing the randomly generated number. This failure was due to the change in `get_country_code_and_national_number` which relied solely on the `national_number` returned from the library, rather than our custom solution to take everything except the country code. **This is a change in behavior**, but I'm inclined to trust the library dedicated to parsing phone numbers.

In addition to tests that are easier to reason with, this should shave >2 minutes off of our test run times.

**Comment by joxl** [comment]
> In addition to tests that are easier to reason with, this should shave >2 minutes off of our test run times.

:1st_place_medal:  :beers:

**Review by joxl — APPROVED** [review]
:partying_face:

**Review by millerdev — APPROVED** [review]
